transport/src/main/java/io/netty/channel/ChannelOutboundInvokerCallback.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+/**
+ * Listener which will be notified when an {@link ChannelOutboundInvoker} methods is completed.
+ */
+// TODO: Should ChannelOutboundInvokerCallback be EventExecutor aware and so be able to delegate if needed ?
+public interface ChannelOutboundInvokerCallback {
+
+    // TODO: Should this be here and should we even support cancellation at all ?
+    static boolean setUncancellable(ChannelOutboundInvokerCallback callback) {
+        if (callback instanceof ChannelPromise) {
+            return ((ChannelPromise) callback).setUncancellable();
+        }
+        return true;
+    }
+
+    /**
+     * Called when an {@link ChannelOutboundInvoker} methods completes successfully.
+     */
+    void onSuccess();
+
+     /**
+     * Called when an {@link ChannelOutboundInvoker} methods completes non-successfully.
+      *
+     * @param cause the {@link Throwable} that caused the error.
+     */
+    void onError(Throwable cause);
```

Comment:
The downside to separate methods like this is we cannot use lambdas to implement a callback.

Reply:
I agree but honestly I wouldn't expect users to implement this interface by themselves. If they want to be notified by something they should use `ctx.write(...).addListener(....)`

Reply:
Alternative we could also make it `Consumer<Throwable>` and if the value is `null` it would be a success. Or even `Consumer<Option<Throwable>>`.

Reply:
Something like this? That *could* work.
```java
    /**
     * Called when an {@link ChannelOutboundInvoker} methods completes.
     *
     * @param error an {@link Optional} containing the {@link Throwable} that caused the error if completed non-successfully, otherwise an empty {@link Optional}
     */
    void onComplete(Consumer<Optional<Throwable>> error);
```

Reply:
Yeah... that said I just added some more docs to explain that the user almost never wants to implement this:

https://github.com/netty/netty/pull/11378/commits/96b43a0271656e20f90c1cef9537c859c9d1c6d8

Reply:
How about `BiConsumer<..., ? extends Throwable>`? `CompletableFuture.handle()` does that and it seemed OK? We could also overload so we have two versions - one with two callback methods and the other with one.

Reply:
Actually after thinking about it I even would consider not let users implement this at all and just make it a final class ... This way it would also be clear to users that they need to use `ChannelFutureListener`. WDYT @trustin @chrisvest @kashike ?

Reply:
Sealed interfaces would be useful here. Making it a class places restrictions on how it can be implemented. I'm also in favour of having descriptive interfaces, rather than relying on generic ones like `BiConsumer`.

Reply:
@chrisvest my point was more about that if its a "final class" people be "forced" to use the right api.

Reply:
Adding another cook to the kitchen: I almost never want to know when something is successful, I usually only want to know if there was an error.

Another random thought: What would you think about making this partially implement Flow. Subscriber (i.e. `onComplete()` and `onError(Throwable)`